File: frontend/src/types/backtest.ts
import type { PageResult } from './api';

export interface BacktestCreateRequest {
  strategy_id: number;
  backtest_name: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  single_order_amount: number;
  data_frequency: string;
  fill_mode: string;
  fee_rate: number;
  stamp_tax_rate: number;
  slippage: number;
}

export interface BacktestDataCheckResult {
  ok: boolean;
  message: string;
  suggestion?: string | null;
  technical_detail?: string | null;
  steps?: BacktestValidationStep[];
}

export interface BacktestValidationStep {
  title: string;
  status: string;
  message: string;
  technical_detail?: string | null;
}

export interface BacktestTaskRecord extends BacktestCreateRequest {
  id: number;
  task_id: string;
  strategy_name: string;
  status: string;
  created_at: string;
}

export interface BacktestResultRecord {
  id: number;
  backtest_id: number;
  total_return: number;
  annual_return: number;
  max_drawdown: number;
  win_rate: number;
  trade_count: number;
  buy_count: number;
  sell_count: number;
  profit_loss_ratio: number;
  average_holding_days: number;
  ending_cash: number;
  open_position_count: number;
  open_market_value: number;
  total_fee: number;
  realized_pnl: number;
  final_cash: number;
  created_at: string;
}

export interface BacktestManifestRecord {
  id: number;
  backtest_id: number;
  strategy_file_name: string;
  strategy_code_hash: string;
  strategy_name: string;
  strategy_version: string;
  data_frequency: string;
  fill_mode: string;
  qmt_mode: string;
  qmt_path: string;
  account_id: string;
  data_coverage_snapshot: string;
  universe_summary: string;
  rule_snapshot: string;
  engine_version: string;
  trust_level: string;
  trust_message: string;
  created_at: string;
}

export interface BacktestStrategySnapshotCheck {
  status: string;
  message: string;
  manifest_hash: string;
  latest_code_hash?: string | null;
  matched_run_id?: string | null;
  matched_task_id?: string | null;
  matched_run_status?: string | null;
  matched_started_at?: string | null;
  matched_finished_at?: string | null;
  latest_run_id?: string | null;
  latest_task_id?: string | null;
  latest_run_status?: string | null;
  latest_started_at?: string | null;
  latest_finished_at?: string | null;
  latest_strategy_file_name?: string | null;
  latest_strategy_version?: string | null;
  technical_detail?: string | null;
}

export interface BacktestTradeRecord {
  id: number;
  backtest_id: number;
  symbol: string;
  name: string;
  side: string;
  price: number;
  quantity: number;
  amount: number;
  fee: number;
  trade_time: string;
  reason: string;
  pnl: number;
}

export interface BacktestSignalRecord {
  id: number;
  backtest_id: number;
  signal_time: string;
  symbol: string;
  name: string;
  action: string;
  price: number;
  amount?: number | null;
  reason: string;
  status: string;
  execution_time?: string | null;
  execution_price?: number | null;
  quantity: number;
  skip_reason?: string | null;
  is_auto_exit: number;
  created_at: string;
}

export interface BacktestEquityRecord {
  id: number;
  backtest_id: number;
  trade_date: string;
  equity: number;
  cash: number;
  market_value: number;
  drawdown: number;
}

export interface BacktestLogRecord {
  id: number;
  backtest_id: number;
  level: string;
  message: string;
  technical_detail?: string | null;
  created_at: string;
}

export interface BacktestReport {
  task: BacktestTaskRecord;
  result?: BacktestResultRecord | null;
  manifest?: BacktestManifestRecord | null;
  strategy_snapshot_check?: BacktestStrategySnapshotCheck | null;
  trades: BacktestTradeRecord[];
  signals: BacktestSignalRecord[];
  equity: BacktestEquityRecord[];
  logs: BacktestLogRecord[];
}

export type BacktestTaskPage = PageResult<BacktestTaskRecord>;
export type BacktestTradePage = PageResult<BacktestTradeRecord>;
export type BacktestSignalPage = PageResult<BacktestSignalRecord>;
export type BacktestLogPage = PageResult<BacktestLogRecord>;

File: backend/schemas/backtest.py
from pydantic import BaseModel, Field


class BacktestCreateRequest(BaseModel):
    strategy_id: int
    backtest_name: str = Field(min_length=1)
    start_date: str
    end_date: str
    initial_cash: float = Field(default=1_000_000, gt=0)
    single_order_amount: float = Field(default=100_000, gt=0)
    data_frequency: str = "日K"
    fill_mode: str = "下一日开盘"
    fee_rate: float = Field(default=0.0003, ge=0)
    stamp_tax_rate: float = Field(default=0.001, ge=0)
    slippage: float = Field(default=0.0, ge=0)


class BacktestDataCheckRequest(BaseModel):
    strategy_id: int
    start_date: str
    end_date: str
    data_frequency: str = "日K"
    fill_mode: str | None = None


class BacktestValidationStep(BaseModel):
    title: str
    status: str
    message: str
    technical_detail: str | None = None


class BacktestDataCheckResult(BaseModel):
    ok: bool
    message: str
    suggestion: str | None = None
    technical_detail: str | None = None
    steps: list[BacktestValidationStep] = Field(default_factory=list)


class BacktestTaskRecord(BaseModel):
    id: int
    task_id: str
    backtest_name: str
    strategy_id: int
    strategy_name: str
    start_date: str
    end_date: str
    initial_cash: float
    single_order_amount: float
    data_frequency: str
    fill_mode: str
    fee_rate: float
    stamp_tax_rate: float
    slippage: float
    status: str
    created_at: str


class BacktestResultRecord(BaseModel):
    id: int
    backtest_id: int
    total_return: float
    annual_return: float
    max_drawdown: float
    win_rate: float
    trade_count: int
    buy_count: int = 0
    sell_count: int = 0
    profit_loss_ratio: float
    average_holding_days: float
    ending_cash: float = 0
    open_position_count: int = 0
    open_market_value: float = 0
    total_fee: float = 0
    realized_pnl: float = 0
    final_cash: float
    created_at: str


class BacktestManifestRecord(BaseModel):
    id: int
    backtest_id: int
    strategy_file_name: str
    strategy_code_hash: str
    strategy_name: str
    strategy_version: str
    data_frequency: str
    fill_mode: str
    qmt_mode: str
    qmt_path: str
    account_id: str
    data_coverage_snapshot: str
    universe_summary: str
    rule_snapshot: str
    engine_version: str
    trust_level: str
    trust_message: str
    created_at: str


class BacktestStrategySnapshotCheck(BaseModel):
    status: str
    message: str
    manifest_hash: str = ""
    latest_code_hash: str | None = None
    matched_run_id: str | None = None
    matched_task_id: str | None = None
    matched_run_status: str | None = None
    matched_started_at: str | None = None
    matched_finished_at: str | None = None
    latest_run_id: str | None = None
    latest_task_id: str | None = None
    latest_run_status: str | None = None
    latest_started_at: str | None = None
    latest_finished_at: str | None = None
    latest_strategy_file_name: str | None = None
    latest_strategy_version: str | None = None
    technical_detail: str | None = None


class BacktestTradeRecord(BaseModel):
    id: int
    backtest_id: int
    symbol: str
    name: str
    side: str
    price: float
    quantity: int
    amount: float
    fee: float
    trade_time: str
    reason: str
    pnl: float


class BacktestSignalRecord(BaseModel):
    id: int
    backtest_id: int
    signal_time: str
    symbol: str
    name: str
    action: str
    price: float
    amount: float | None = None
    reason: str
    status: str
    execution_time: str | None = None
    execution_price: float | None = None
    quantity: int = 0
    skip_reason: str | None = None
    is_auto_exit: int = 0
    created_at: str


class BacktestEquityRecord(BaseModel):
    id: int
    backtest_id: int
    trade_date: str
    equity: float
    cash: float
    market_value: float
    drawdown: float


class BacktestLogRecord(BaseModel):
    id: int
    backtest_id: int
    level: str
    message: str
    technical_detail: str | None = None
    created_at: str


class BacktestReport(BaseModel):
    task: BacktestTaskRecord
    result: BacktestResultRecord | None = None
    manifest: BacktestManifestRecord | None = None
    strategy_snapshot_check: BacktestStrategySnapshotCheck | None = None
    trades: list[BacktestTradeRecord]
    signals: list[BacktestSignalRecord] = []
    equity: list[BacktestEquityRecord]
    logs: list[BacktestLogRecord]
